fuse/fuse_test.go

```diff
@@ -480,3 +480,99 @@ func TestWithRoot(t *testing.T) {
+	BenchmarkDataSizes = []int64{
+		0,
+		1024, 2*1024, 16*1024, 64*1024, 128*1023,
+		1*1024*1024, 16*1024*1024,
+	}
+)
+
+func stageAndRead(b *testing.B, ctx context.Context, control *spawntest.Control, mount *mountInfo, label string, data []byte) {
+	size := len(data)
+	// stage data to catFS
+	catfsFilePath := fmt.Sprintf("%s_file_%d", label, size)
+	req := catfsPayload{Path: catfsFilePath, Data: data}
+	require.Nil(b, control.JSON("/catfsStage").Call(ctx, req, &nothing{}))
+	fuseFilePath := filepath.Join(mount.Dir, catfsFilePath)
+
+	// Read it back via fuse
+	b.Run(fmt.Sprintf("%s_Size_%d", label, size), func(b *testing.B) {
+		for n := 0; n < b.N; n++ {
+			ioutil.ReadFile(fuseFilePath)
+		}
+	})
+}
+
+func BenchmarkRead(b *testing.B) {
+	withMount(b, MountOptions{}, func(ctx context.Context, control *spawntest.Control, mount *mountInfo) {
+		for _, size := range BenchmarkDataSizes {
+			// Check how fast is readout of a file with compressible content
+			data := testutil.CreateDummyBuf(size)
+			stageAndRead(b, ctx, control, mount,  "CompressibleContent", data)
+
+			// Check how fast is readout of a file with random/uncompressible content
+			data = testutil.CreateRandomDummyBuf(size, 1)
+			stageAndRead(b, ctx, control, mount,  "RandomContent", data)
+		}
+	})
+}
+
+func writeDataNtimes(b *testing.B, data []byte, ntimes int) {
+	// Writing could be very space demanding even for a small size,
+	// Since benchmark runs many-many times it will consume a lot of space
+	// we have to remount everything every time to start with clean catFS DB
+	// Consequently, this test takes long time, since mounting is long operation
+	label := "dummy"
+	size := len(data)
+	for n := 0; n < b.N; n++ {
+		b.StopTimer()
+		withMount(b, MountOptions{}, func(ctx context.Context, control *spawntest.Control, mount *mountInfo) {
+			// Check how fast is write to a file with compressible content
+			catfsFilePath := fmt.Sprintf("%s_file_%d", label, size)
+			fuseFilePath := filepath.Join(mount.Dir, catfsFilePath)
+
+			b.StartTimer()
+			for i := 0; i < ntimes; i++ {
+				if len(data)>0 {
+					// generate new content for backend
+					data[0]++
+				}
+			}
+			require.Nil(b, ioutil.WriteFile(fuseFilePath, data, 0644))
+			b.StopTimer()
+		})
+	}
+}
+
+var (
+	// keep this low or you might run out of space
+	NumberOfOverWrites = []int{
+		1, 2,  5,
+	}
+)
+
+func BenchmarkWrite(b *testing.B) {
+	size := int64(10*1024*1024)
+
+	for _, Ntimes := range NumberOfOverWrites {
+		// Check how fast is write to a file with compressible content
+		data := testutil.CreateDummyBuf(size)
+		prefix := fmt.Sprintf("Owerwrite_%d", Ntimes)
+		label := fmt.Sprintf("%s/CompressibleContent_Size_%d", prefix, size)
+		b.Run(label, func(b *testing.B) {
+			writeDataNtimes(b, data, Ntimes)
+		})
+
+		// Check how fast is write to a file with random/uncompressible content
+		data = testutil.CreateRandomDummyBuf(size, 1)
+		label = fmt.Sprintf("%s/RandomContent_Size_%d", prefix, size)
```

Comment:
Random content is about factor of 3 slower to read and write. Not sure what to blame. Is it because backend stores or serve less data? Why is it factor of 3?

Reply:
I guess it's because it's way more data that needs to be encrypted / decrypted. But that's the reason why we need also benchmarks for parts of the pipeline.